tests/garage/torch/modules/test_cnn_module.py

```diff
@@ -0,0 +1,183 @@
+"""Test CNNModule."""
+import pickle
+
+import numpy as np
+import pytest
+import torch
+import torch.nn as nn
+
+from garage.torch.modules import CNNModule
+
+
+class TestCNNModule:
+    """Test CNNModule."""
+
+    def setup_method(self):
+        self.batch_size = 64
+        self.input_width = 32
+        self.input_height = 32
+        self.in_channel = 3
+        self.dtype = torch.float32
+        self.input = torch.zeros(
+            (self.batch_size, self.in_channel, self.input_height,
+             self.input_width),
+            dtype=self.dtype)  # minibatch size 64, image size [3, 32, 32]
+
+    @pytest.mark.parametrize(
+        'kernel_sizes, hidden_channels, strides, paddings', [
+            ((1, ), (32, ), (1, ), (0, )),
+            ((3, ), (32, ), (1, ), (0, )),
+            ((3, ), (32, ), (2, ), (0, )),
+            ((5, ), (12, ), (1, ), (2, )),
+            ((1, 1), (32, 64), (1, 1), (0, 0)),
+            ((3, 3), (32, 64), (1, 1), (0, 0)),
+            ((3, 3), (32, 64), (2, 2), (0, 0)),
+        ])
+    def test_output_values(self, kernel_sizes, hidden_channels, strides,
+                           paddings):
+        """Test output values from CNNBaseModule.
+
+        Args:
+            kernel_sizes (tuple[int]): Kernel sizes.
+            hidden_channels (tuple[int]): hidden channels.
+            strides (tuple[int]): strides.
+            paddings (tuple[int]): value of zero-padding.
+
+        """
+        module_with_nonlinear_function_and_module = CNNModule(
+            input_var=self.input,
+            hidden_channels=hidden_channels,
+            kernel_sizes=kernel_sizes,
+            strides=strides,
+            paddings=paddings,
+            padding_mode='zeros',
+            hidden_nonlinearity=torch.relu,
+            hidden_w_init=nn.init.xavier_uniform_)
+
+        module_with_nonlinear_module_instance_and_function = CNNModule(
+            input_var=self.input,
+            hidden_channels=hidden_channels,
+            kernel_sizes=kernel_sizes,
+            strides=strides,
+            paddings=paddings,
+            padding_mode='zeros',
+            hidden_nonlinearity=nn.ReLU(),
+            hidden_w_init=nn.init.xavier_uniform_)
+
+        output1 = module_with_nonlinear_function_and_module(self.input)
+        output2 = module_with_nonlinear_module_instance_and_function(
+            self.input)
+
+        current_size = self.input_width
+        for (filter_size, stride, padding) in zip(kernel_sizes, strides,
+                                                  paddings):
+            # padding = float((filter_size - 1) / 2) # P = (F - 1) /2
+            current_size = int(
+                (current_size - filter_size + padding * 2) /
+                stride) + 1  # conv formula = ((W - F + 2P) / S) + 1
+        flatten_shape = current_size * current_size * hidden_channels[-1]
+
+        assert output1.shape == (self.batch_size, flatten_shape)
```

Comment:
can you test values here, and not just shapes? see the TF tree tests for some examples you can probably copy.